beps/0003-metadata-on-relations/README.md

```diff
@@ -0,0 +1,358 @@
+
+## Dependencies
+
+<!--
+List any dependencies that this work has on other BEPs or features.
+-->
+
+None that we know of.
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+## 1. Supplying override data on the Component itself.
+
+This approach would look something like,
+
+```yaml
+kind: Component
+spec:
+  environmentOverrides:
+    production:
+      key1: value1
+    development:
+      key1: value2
+```
+
+The primary downside of this approach is that this is a fragile API that doesn't scale well to additional use cases. There's no easy way to add a new override for an entirely different context. Either you have to start nesting keys,
+
+```yaml
+overrides:
+  environment:
+    production: ...
+  otherOverride:
+    key2: ...
+```
+
+or you have to start defining top level keys,
+
+```yaml
+environmentOverrides:
+  production: ...
+otherOverride:
+  key2: ...
+```
+
+And both of those don't scale well with complexity. We also have the connection between the two entities we care about, and in all cases there are multiple entities involved. That connection should contain the data, not an entity participating in that connection.
+
+### 2. Relations (with metadata) as separate objects
+
+This is something that was floated as part of [#1964](https://github.com/backstage/backstage/issues/1964) as well as came up in [the Environment kind PRFC](https://github.com/backstage/backstage/pull/21560#issuecomment-1826217370). The idea being that because these relations are starting to resemble entities with metadata, that they should be stored as such in that catalog instead of us just attaching metadata to them and keeping them as relations. This is a more common pattern in Kubernetes and not something we've seen in Backstage. `@Rugvip`'s comment there is still relevant,
+
+> The suggested solution also doesn't model relations as standalone entities. This is primarily due to the overhead of this approach, especially since we'd likely end up having to name each relation individually. It may make sense for some larger and more complex relations as separate entities though, and could be combined with the suggestion in this RFC. It does feel very overkill for relations like owner and dependency though.
+> I'm also a bit worried that modelling relations as entities may cause confusion as they extend the existing entity model. What are for example the role of labels in relation entities, would you query for a set of relations using a label query? ^\_> It's also not as clear how to map that model to GraphQL IMO.
+
+Not only would relations as entities introduce a bunch of complexity to the catalog, but it would still be overkill for our use case. We would argue that `consumesApi` and `deploysTo` are not much different than `ownerOf` or `dependencyOf`.
```

Comment:
I see ‘consumesApi’ as very similar to ‘dependencyOf’ indeed (e.g. we’d love to use this feature to mark optional or conditional dependencies). It might also be useful to define the role of a User in a Group and similar simple use-cases. I’d love to show such info in tooltips in the diagrams, or as a column in a table showing relationships. And yep, “relationships as entities” seems like an overkill for that.

On the other hand I perceive ‘deploysTo’ more as a composition of two relations (Component deployedAs ComponentDeployment, ComponentDeployment partOf Environment - hopefully naming is clear enough), where the interesting metadata would be part of the ComponentDeployment entity. Some reasons:
 - Component and ComponentDeployment can have different owners (in fact, different ComponentDeployments for the same Component could have different owners)
 - I’d expect information about Components to be defined manually, but information about deployments to be retrieved from some system with a different entity provider
 - We use this approach for representing deployments, and to react on @Rugvip, we actually do filter the ComponentDeployment entities by labels (or owners)